redaxo/src/addons/media_manager/lib/effects/effect_img2img.php

```diff
@@ -0,0 +1,119 @@
+<?php
+
+/**
+ * @package redaxo\media-manager
+ */
+
+class rex_effect_img2img extends rex_effect_abstract
+{
+    private static $convert_types = [
+        'jpg',
+        'png',
+        'gif',
+        'webp',
+    ];
+
+    private static $convert_to = [
+        'jpg' => [
+            'ext' => 'jpg',
+            'content-type' => 'image/jpeg',
+        ],
+        'png' => [
+            'ext' => 'png',
+            'content-type' => 'image/png',
+        ],
+        'gif' => [
+            'ext' => 'gif',
+            'content-type' => 'image/gif',
+        ],
+        'webp' => [
+            'ext' => 'webp',
+            'content-type' => 'image/webp',
+        ],
+    ];
+
+    private static $convert_tos = ['jpg', 'png', 'gif', 'webp'];
+    private static $convert_to_default = 'jpg';
+
+    public function execute()
+    {
+        $media = $this->media;
+
+        $media->asImage();
+        $imageObject = $media->getImage();
+
+        $this->keepTransparent($imageObject);
+
+        if (null === $imageObject) {
+            return;
+        }
+
+        $ext = $media->getFormat();
+        // skip if extension is not in list
+        if (!in_array(strtolower($ext), self::$convert_types)) {
+            return;
+        }
+
+        if (!isset(self::$convert_to[$this->params['convert_to']])) {
+            $convert_to = self::$convert_to[self::$convert_to_default];
+        } else {
+            $convert_to = self::$convert_to[$this->params['convert_to']];
+        }
+
+        // convert image
+        $addon = rex_addon::get('media_manager');
+
+        imagepalettetotruecolor($imageObject);
+
+        $interlace = $media->getImageProperty(rex_managed_media::PROP_INTERLACE, $addon->getConfig('interlace'));
+        imageinterlace($imageObject, in_array($convert_to['ext'], $interlace) ? 1 : 0);
+
+        switch ($convert_to['ext']) {
+            case 'jpg':
+                $quality = $media->getImageProperty(rex_managed_media::PROP_JPG_QUALITY, $addon->getConfig('jpg_quality'));
+                imagejpeg($imageObject, null, $quality);
```

Comment:
Die Doku sagt zu dem zweiten Parameter:

> Der Pfad unter dem das Bild gespeichert werden soll. Ist dies nicht gesetzt oder null wird der rohe Bilddatenstrom direkt ausgegeben.

https://www.php.net/manual/de/function.imagejpeg.php

Wenn ich das richtig sehe, ist daher der Ablauf aktuell so:

- Hier wird das Bild direkt ausgegeben (wenn ich das richtig verstehe, wird aber nix an der $imageObject-Resource geändert)
- Es läuft aber der Output-Buffer von REDAXO
- Später im Prozess [leert der MM den Output-Buffer](https://github.com/redaxo/redaxo/blob/3b1133b71e548eeb10f5c3ca4354d4797027e700/redaxo/src/addons/media_manager/lib/managed_media.php#L174), die Ausgabe von hier geht also verloren
- Und der MM gibt das Bild am Ende erneut aus
- Da kommt trotzdem das richtig Format raus, da dieser Effekt ja das Format per `setFormat` ändert, und deswegen [nimmt dann der MM am Ende erneut die korrekte `image*`-Methode](https://github.com/redaxo/redaxo/blob/3b1133b71e548eeb10f5c3ca4354d4797027e700/redaxo/src/addons/media_manager/lib/managed_media.php#L232) (die gleich, wie zuvor hier im Effekt)

Oder mit anderen Worten: Die Aufrufe hier scheinen mir unnötig zu sein und im aktuellen Ablauf auch keine Auswirkung auf das Bild am Ende zu haben.

Reply:
@gharlan also hier rausnehmen und durch den Fix im image-properties + Setzen des Formats hier macht 
der MM dann alles richtig !?

Reply:
Ja, mir scheint es so. Also scheinbar reicht es für diesen Effekt, wenn er `asImage()` aufruft, und `setFormat`, `setMediaFilename` und `setHeader`.
`asImage` sollte er aufrufen, aus dem gleichen Grund wie in meinem anderen PR, also im die gd-resource zu erzwingen. denn nur dann wird es am Ende über die `image*`-Funktion abgelegt.

Reply:
@gharlan ok, habe das jetzt soweit fertig und update den PR heute Abend.

Kleiner Fix für WEBPist notwendig (PNG mit indizierten Farben)

Bei GIF ist auch eine Anpassung notwendig. 
Hier gibt es Probleme mit der Transparenz. Man kann ja entweder mit schwarz für dunkle Hintergründe
oder weiß für helle Hintergründe arbeiten. Seht ihr es als notwendig hier Schwarz/Weiss als Parameter
vorzugeben bei GIF!? @gharlan @alxndr-w @skerbis 

Am meisten Anwendung wird ja m.E. die WEBP-Umwandlung haben.

Reply:
- Eine Umwandlung zu gif finde ich eher unnötig. 
- Umwandlung von GIF zu den anderen, da wäre es schon gut einen Farbwert angeben zu können.

Reply:
Ich würde nicht pauschal (für alle potentiellen Anwender) entscheiden, was sinnvoll ist und was nicht. Das Ding sollte alles in alles konvertieren können (aus dem Set der aktuell 4 Typen). Es wird 1, 2 Fälle geben, die 95% ausmachen werden, aber sei's drum.

> Bei GIF ist auch eine Anpassung notwendig.
> Hier gibt es Probleme mit der Transparenz. Man kann ja entweder mit schwarz für dunkle Hintergründe

Gibt es nicht schon 1 Effekt der die Basisfläche einfärbt? Falls ja, würden wir da nicht schon wieder mischen? Wenn es so einen Effekt gibt und der Fall GIF > JPG (als Beispiel) auftaucht, muss der Mensch, der das MM-Set erstellt, das eben berücksichtigen und besagten Effekt noch davorschalten.

Reply:
![image](https://user-images.githubusercontent.com/330560/108249284-2c0adb80-7155-11eb-8031-272a3a30f75c.png)

Es geht hier um die Transparenz (Übergänge) und nicht um die Hintergrundfarbe
Zur Verdeutlichung der Screenshot. Links mit weiss und rechts mit schwarz.

Bei dunklem Hintergrund ist klar schwarz besser. Bei hellem Hintergrund eben weiss.

Reply:
Der hier schwarze Bereich ist transparent!

Reply:
![image](https://user-images.githubusercontent.com/330560/108249924-f4506380-7155-11eb-9bc3-2ef727f12a02.png)

Reply:
Verstehe ich nicht. Gibt es bei webp keine Alpha-Transparenz, sondern nur wie bei GIF binäre Transparenz?

Reply:
hier geht es um gif und nicht um webp :)

Reply:
ah jetzt wird es klarer. Wenn sich das einbauen lässt wäre ich dafür.